File: src/TestApp/DashboardSnapshotFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MyServiceBus;
using MyServiceBus.Inspection;

namespace TestApp;

public static class DashboardSnapshotFactory
{
    public static DashboardOverviewDto CreateOverview(IMessageBus bus, DashboardMetadata metadata)
        => CreateOverview(new BusInspectionProvider(bus).GetSnapshot(), metadata);

    public static DashboardOverviewDto CreateOverview(IMessageBus bus, DashboardMetadata metadata, DashboardState state)
        => CreateOverview(new BusInspectionProvider(bus).GetSnapshot(), metadata, state.IsStarted ? "Healthy" : "Unknown");

    public static DashboardOverviewDto CreateOverview(BusInspectionSnapshot snapshot, DashboardMetadata metadata)
        => new(
            metadata.ServiceName,
            metadata.TransportName,
            snapshot.Address.ToString(),
            snapshot.CapturedAt.UtcDateTime,
            snapshot.Messages.Count,
            snapshot.Consumers.Count,
            snapshot.ReceiveEndpoints.Count,
            new DashboardHealthLinksDto("/health/ready", "/health/live", null));

    public static DashboardOverviewDto CreateOverview(BusInspectionSnapshot snapshot, DashboardMetadata metadata, string? status)
        => new(
            metadata.ServiceName,
            metadata.TransportName,
            snapshot.Address.ToString(),
            snapshot.CapturedAt.UtcDateTime,
            snapshot.Messages.Count,
            snapshot.Consumers.Count,
            snapshot.ReceiveEndpoints.Count,
            new DashboardHealthLinksDto("/health/ready", "/health/live", status));

    public static DashboardMessagesDto CreateMessages(IMessageBus bus, DashboardMetadata metadata)
        => CreateMessages(new BusInspectionProvider(bus).GetSnapshot(), metadata);

    public static DashboardMessagesDto CreateMessages(BusInspectionSnapshot snapshot, DashboardMetadata metadata)
        => new(
            metadata.ServiceName,
            metadata.TransportName,
            snapshot.Address.ToString(),
            snapshot.CapturedAt.UtcDateTime,
            snapshot.Messages
                .Select(x => new DashboardMessageDto(x.MessageType, x.MessageUrn, x.EntityName, x.ImplementedMessageTypes))
                .ToArray());

    public static DashboardConsumersDto CreateConsumers(IMessageBus bus, DashboardMetadata metadata)
        => CreateConsumers(new BusInspectionProvider(bus).GetSnapshot(), metadata);

    public static DashboardConsumersDto CreateConsumers(BusInspectionSnapshot snapshot, DashboardMetadata metadata)
        => new(
            metadata.ServiceName,
            metadata.TransportName,
            snapshot.Address.ToString(),
            snapshot.CapturedAt.UtcDateTime,
            snapshot.Consumers
                .Select(x => new DashboardConsumerDto(
                    x.ConsumerType,
                    x.EndpointName,
                    x.PrefetchCount,
                    x.SerializerType,
                    x.Properties,
                    snapshot.ReceiveEndpoints.First(e => e.EndpointName == x.EndpointName).Bindings
                        .Select(b => new DashboardBindingDto(b.MessageType, b.MessageUrn, b.EntityName))
                        .ToArray()))
                .ToArray());

    public static DashboardTopologyDto CreateTopology(IMessageBus bus, DashboardMetadata metadata)
        => CreateTopology(new BusInspectionProvider(bus).GetSnapshot(), metadata);

    public static DashboardTopologyDto CreateTopology(BusInspectionSnapshot snapshot, DashboardMetadata metadata)
        => new(
            metadata.ServiceName,
            metadata.TransportName,
            snapshot.Address.ToString(),
            snapshot.CapturedAt.UtcDateTime,
            CreateMessages(snapshot, metadata).Messages,
            CreateConsumers(snapshot, metadata).Consumers);

    public static DashboardQueuesDto CreateQueues(IMessageBus bus, DashboardMetadata metadata)
        => CreateQueues(new BusInspectionProvider(bus).GetSnapshot(), metadata);

    public static DashboardQueuesDto CreateQueues(BusInspectionSnapshot snapshot, DashboardMetadata metadata)
        => new(
            metadata.ServiceName,
            metadata.TransportName,
            snapshot.Address.ToString(),
            snapshot.CapturedAt.UtcDateTime,
            snapshot.ReceiveEndpoints
                .Select(x => new DashboardQueueDto(
                    x.EndpointName,
                    Durable: x.Transport?.Properties.TryGetValue("durable", out var durable) == true && durable is bool durableValue ? durableValue : true,
                    AutoDelete: x.Transport?.Properties.TryGetValue("autoDelete", out var autoDelete) == true && autoDelete is bool autoDeleteValue ? autoDeleteValue : false,
                    PrefetchCount: snapshot.Consumers.FirstOrDefault(c => c.EndpointName == x.EndpointName)?.PrefetchCount,
                    QueueArguments: x.Properties,
                    Consumers: x.ConsumerTypes,
                    Bindings: x.Bindings.Select(b => new DashboardBindingDto(b.MessageType, b.MessageUrn, b.EntityName)).ToArray(),
                    ErrorQueueName: x.Transport?.Properties.TryGetValue("errorQueueName", out var errorQueueName) == true ? errorQueueName?.ToString() ?? $"{x.EndpointName}_error" : $"{x.EndpointName}_error",
                    FaultQueueName: x.Transport?.Properties.TryGetValue("faultQueueName", out var faultQueueName) == true ? faultQueueName?.ToString() ?? $"{x.EndpointName}_fault" : $"{x.EndpointName}_fault",
                    SkippedQueueName: x.Transport?.Properties.TryGetValue("skippedQueueName", out var skippedQueueName) == true ? skippedQueueName?.ToString() ?? $"{x.EndpointName}_skipped" : $"{x.EndpointName}_skipped"))
                .ToArray());

    public static DashboardMetricsDto CreateMetrics(IMessageBus bus, DashboardMetadata metadata, DashboardState state)
    {
        var generatedAtUtc = DateTime.UtcNow;
        var snapshot = state.CreateMetricsSnapshot(generatedAtUtc);
        return new DashboardMetricsDto(
            metadata.ServiceName,
            metadata.TransportName,
            bus.Address.ToString(),
            generatedAtUtc,
            snapshot.Totals,
            snapshot.Rates,
            snapshot.Latency,
            snapshot.Queues,
            snapshot.Messages);
    }

    public static string? ResolveMessageTypeFromDestination(Uri destinationAddress)
    {
        var segments = destinationAddress.AbsolutePath.Trim('/').Split('/', StringSplitOptions.RemoveEmptyEntries);
        if (segments.Length == 0)
            return null;

        var finalSegment = segments[^1];
        return string.IsNullOrWhiteSpace(finalSegment) ? null : finalSegment;
    }

    public static string? ResolveMessageUrnFromDestination(Uri destinationAddress)
    {
        var messageType = ResolveMessageTypeFromDestination(destinationAddress);
        return messageType is null ? null : $"destination:{messageType}";
    }
}

File: src/Java/testapp/src/main/java/com/myservicebus/testapp/dashboard/DashboardSnapshotFactory.java
package com.myservicebus.testapp.dashboard;

import com.myservicebus.MessageUrn;
import com.myservicebus.inspection.BusInspectionProvider;
import com.myservicebus.inspection.BusInspectionSnapshot;

import java.time.Instant;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class DashboardSnapshotFactory {
    private DashboardSnapshotFactory() {
    }

    public static DashboardOverviewDto createOverview(BusInspectionProvider inspectionProvider, DashboardMetadata metadata) {
        return createOverview(inspectionProvider.getSnapshot(), metadata);
    }

    public static DashboardOverviewDto createOverview(BusInspectionProvider inspectionProvider, DashboardMetadata metadata, DashboardState state) {
        return createOverview(inspectionProvider.getSnapshot(), metadata, state.isStarted() ? "Healthy" : "Unknown");
    }

    public static DashboardMessagesDto createMessages(BusInspectionProvider inspectionProvider, DashboardMetadata metadata) {
        return createMessages(inspectionProvider.getSnapshot(), metadata);
    }

    public static DashboardOverviewDto createOverview(BusInspectionSnapshot snapshot, DashboardMetadata metadata) {
        return new DashboardOverviewDto(
                metadata.serviceName(),
                metadata.transportName(),
                snapshot.address().toString(),
                snapshot.capturedAt(),
                snapshot.messages().size(),
                snapshot.consumers().size(),
                snapshot.receiveEndpoints().size(),
                new DashboardHealthLinksDto("/health/ready", "/health/live", null));
    }

    public static DashboardOverviewDto createOverview(BusInspectionSnapshot snapshot, DashboardMetadata metadata, String status) {
        return new DashboardOverviewDto(
                metadata.serviceName(),
                metadata.transportName(),
                snapshot.address().toString(),
                snapshot.capturedAt(),
                snapshot.messages().size(),
                snapshot.consumers().size(),
                snapshot.receiveEndpoints().size(),
                new DashboardHealthLinksDto("/health/ready", "/health/live", status));
    }

    public static DashboardMessagesDto createMessages(BusInspectionSnapshot snapshot, DashboardMetadata metadata) {
        return new DashboardMessagesDto(
                metadata.serviceName(),
                metadata.transportName(),
                snapshot.address().toString(),
                snapshot.capturedAt(),
                snapshot.messages().stream()
                        .map(x -> new DashboardMessageDto(x.messageType(), x.messageUrn(), x.entityName(), x.implementedMessageTypes()))
                        .toList());
    }

    public static DashboardConsumersDto createConsumers(BusInspectionProvider inspectionProvider, DashboardMetadata metadata) {
        return createConsumers(inspectionProvider.getSnapshot(), metadata);
    }

    public static DashboardConsumersDto createConsumers(BusInspectionSnapshot snapshot, DashboardMetadata metadata) {
        return new DashboardConsumersDto(
                metadata.serviceName(),
                metadata.transportName(),
                snapshot.address().toString(),
                snapshot.capturedAt(),
                snapshot.consumers().stream()
                        .map(x -> new DashboardConsumerDto(
                                x.consumerType(),
                                x.endpointName(),
                                x.prefetchCount(),
                                x.serializerType(),
                                x.properties(),
                                snapshot.receiveEndpoints().stream()
                                        .filter(e -> e.endpointName().equals(x.endpointName()))
                                        .findFirst()
                                        .map(e -> e.bindings().stream()
                                                .map(b -> new DashboardBindingDto(b.messageType(), b.messageUrn(), b.entityName()))
                                                .toList())
                                        .orElse(List.of())))
                        .toList());
    }

    public static DashboardTopologyDto createTopology(BusInspectionProvider inspectionProvider, DashboardMetadata metadata) {
        return createTopology(inspectionProvider.getSnapshot(), metadata);
    }

    public static DashboardTopologyDto createTopology(BusInspectionSnapshot snapshot, DashboardMetadata metadata) {
        return new DashboardTopologyDto(
                metadata.serviceName(),
                metadata.transportName(),
                snapshot.address().toString(),
                snapshot.capturedAt(),
                createMessages(snapshot, metadata).messages(),
                createConsumers(snapshot, metadata).consumers());
    }

    public static DashboardQueuesDto createQueues(BusInspectionProvider inspectionProvider, DashboardMetadata metadata) {
        return createQueues(inspectionProvider.getSnapshot(), metadata);
    }

    public static DashboardQueuesDto createQueues(BusInspectionSnapshot snapshot, DashboardMetadata metadata) {
        return new DashboardQueuesDto(
                metadata.serviceName(),
                metadata.transportName(),
                snapshot.address().toString(),
                snapshot.capturedAt(),
                snapshot.receiveEndpoints().stream()
                        .map(x -> new DashboardQueueDto(
                                x.endpointName(),
                                x.transport() != null && Boolean.TRUE.equals(x.transport().properties().get("durable")),
                                x.transport() != null && Boolean.TRUE.equals(x.transport().properties().get("autoDelete")),
                                snapshot.consumers().stream()
                                        .filter(c -> c.endpointName().equals(x.endpointName()))
                                        .findFirst()
                                        .map(BusInspectionSnapshot.ConsumerInspection::prefetchCount)
                                        .orElse(null),
                                x.properties(),
                                x.consumerTypes(),
                                x.bindings().stream()
                                        .map(b -> new DashboardBindingDto(b.messageType(), b.messageUrn(), b.entityName()))
                                        .toList(),
                                x.transport() != null && x.transport().properties().containsKey("errorQueueName") ? x.transport().properties().get("errorQueueName").toString() : x.endpointName() + "_error",
                                x.transport() != null && x.transport().properties().containsKey("faultQueueName") ? x.transport().properties().get("faultQueueName").toString() : x.endpointName() + "_fault",
                                x.transport() != null && x.transport().properties().containsKey("skippedQueueName") ? x.transport().properties().get("skippedQueueName").toString() : x.endpointName() + "_skipped"))
                        .toList());
    }

    public static DashboardMetricsDto createMetrics(BusInspectionProvider inspectionProvider, DashboardMetadata metadata, DashboardState state) {
        Instant generatedAtUtc = Instant.now();
        BusInspectionSnapshot inspectionSnapshot = inspectionProvider.getSnapshot();
        DashboardState.DashboardMetricsSnapshot snapshot = state.createMetricsSnapshot(generatedAtUtc);
        return new DashboardMetricsDto(
                metadata.serviceName(),
                metadata.transportName(),
                inspectionSnapshot.address().toString(),
                generatedAtUtc,
                snapshot.totals(),
                snapshot.rates(),
                snapshot.latency(),
                snapshot.queues(),
                snapshot.messages());
    }

    public static String resolveMessageTypeFromDestination(java.net.URI destinationAddress) {
        String path = destinationAddress.getPath();
        if (path == null || path.isBlank()) {
            return null;
        }

        String[] segments = path.replaceAll("^/+", "").split("/");
        if (segments.length == 0) {
            return null;
        }

        String finalSegment = segments[segments.length - 1];
        return finalSegment.isBlank() ? null : finalSegment;
    }

    public static String resolveMessageUrnFromDestination(java.net.URI destinationAddress) {
        String messageType = resolveMessageTypeFromDestination(destinationAddress);
        return messageType == null ? null : "destination:" + messageType;
    }

    public record DashboardOverviewDto(
            String serviceName,
            String transportName,
            String address,
            Instant generatedAtUtc,
            int messageCount,
            int consumerCount,
            int queueCount,
            DashboardHealthLinksDto health) {
    }

    public record DashboardMessagesDto(
            String serviceName,
            String transportName,
            String address,
            Instant generatedAtUtc,
            List<DashboardMessageDto> messages) {
    }

    public record DashboardConsumersDto(
            String serviceName,
            String transportName,
            String address,
            Instant generatedAtUtc,
            List<DashboardConsumerDto> consumers) {
    }

    public record DashboardTopologyDto(
            String serviceName,
            String transportName,
            String address,
            Instant generatedAtUtc,
            List<DashboardMessageDto> messages,
            List<DashboardConsumerDto> consumers) {
    }

    public record DashboardQueuesDto(
            String serviceName,
            String transportName,
            String address,
            Instant generatedAtUtc,
            List<DashboardQueueDto> queues) {
    }

    public record DashboardMetricsDto(
            String serviceName,
            String transportName,
            String address,
            Instant generatedAtUtc,
            DashboardCounterSetDto totals,
            DashboardRateSetDto rates,
            DashboardLatencySetDto latency,
            List<DashboardQueueMetricsDto> byQueue,
            List<DashboardMessageMetricsDto> byMessageType) {
    }

    public record DashboardMessageDto(
            String messageType,
            String messageUrn,
            String entityName,
            List<String> implementedMessageTypes) {
    }

    public record DashboardConsumerDto(
            String consumerType,
            String queueName,
            Integer prefetchCount,
            String serializerType,
            Map<String, Object> queueArguments,
            List<DashboardBindingDto> bindings) {
    }

    public record DashboardBindingDto(
            String messageType,
            String messageUrn,
            String entityName) {
    }

    public record DashboardQueueDto(
            String queueName,
            boolean durable,
            boolean autoDelete,
            Integer prefetchCount,
            Map<String, Object> queueArguments,
            List<String> consumers,
            List<DashboardBindingDto> bindings,
            String errorQueueName,
            String faultQueueName,
            String skippedQueueName) {
    }

    public record DashboardHealthLinksDto(
            String readyUrl,
            String liveUrl,
            String status) {
    }

    public record DashboardCounterSetDto(
            long published,
            long sent,
            long consumed,
            long faulted) {
    }

    public record DashboardRateSetDto(
            double publishedPerSecond,
            double sentPerSecond,
            double consumedPerSecond) {
    }

    public record DashboardLatencySetDto(
            Double consumeAvgMs) {
    }

    public record DashboardQueueMetricsDto(
            String queueName,
            long consumed,
            long faulted,
            long inFlight) {
    }

    public record DashboardMessageMetricsDto(
            String messageType,
            String messageUrn,
            long published,
            long sent,
            long consumed,
            long faulted) {
    }
}
